packages/libs/deer/src/lib.rs

```diff
@@ -1,9 +1,451 @@
+// `deserialize_u*` methods.
+struct NumberVisitor<E: Error>(PhantomData<fn() -> E>);
+
+impl<E: Error> NumberVisitor<E> {
+    fn new() -> Self {
+        Self(PhantomData::default())
+    }
+}
+
+impl<E: Error> Visitor<'_> for NumberVisitor<E> {
+    type Error = E;
+    type Value = Number;
+
+    fn expecting_display(&self) -> String {
+        "number".to_string()
+    }
+
+    fn visit_number(self, v: Number) -> Result<Self::Value, Self::Error> {
+        Ok(v)
+    }
+}
+
+macro_rules! derive_from_number {
+    [$($method:ident ($to:ident) -> $visit:ident,)*] => {
+        $(derive_from_number!(#internal, $method; $to, $visit);)*
+    };
+
+    (#internal, $method:ident; $to:ident, $visit:ident) => {
+        /// Automatically implemented convenience method, which uses [`Self::deserialize_number`]
+        /// to extract a value of the primitive type, will otherwise error out.
+        ///
+        /// # Errors
+        ///
+        /// Current value is either not a number or wasn't able to be casted to the primitive type
+        fn $method<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+        where
+            V: Visitor<'de>,
+        {
+            let n = self.deserialize_number(NumberVisitor::<Self::Error>::new())?;
+            let v = n
+                .$to()
+                .ok_or_else(|| Self::Error::message("provided value too large or too small"))
+                .into_report()?;
+
+            visitor.$visit(v).change_context(Self::Error::new())
+        }
+    };
+}
+
+/// A **data format** that can deserialize any data structure which is supported by deer.
+///
+/// This traits defines the deserialization half, while [`Deserialize`] is used to use the
+/// deserializer to convert that data into the instance of a type, or fail if that was not possible.
+///
+/// `deer`s focus is on simplicity, therefore the data model is a lot less specific than ones used
+/// by other crates (like [`serde`]).
+///
+/// The data model consists of the following types:
+///
+/// * 6 primitives:
+///     * `null`
+///         * encodes the explicit absence of a value (`undefined`/`null` in JSON)
+///     * `bool`
+///         * Rust equivalent: [`true`], [`false`]
+///     * `number`
+///         * encodes both floating point and integral numbers
+///     * `char`:
+///         * example: `'a'`
+///     * `string`
+///         * example: `"Hello World!"`
+///     * `bytes`
+///         * example: `[0b0001, 0b1000]`
+/// * composite types
+///     * `object`
+///         * encodes any object, be it a map, struct or enum struct variant
+///     * `array`
+///         * encodes any sequence of data, be it an array, a set or tuple
+///
+/// The [`Deserializer`] trait supports a single entrypoint, which are methods that consume the
+/// [`Deserializer`] and either return the value requested or return an error.
+///
+/// [`serde`]: https://serde.rs/
+pub trait Deserializer<'de>: Sized {
```

Comment:
This seems fairly comprehensive and perhaps bigger than originally envisioned? I am rather concerned that we're going to fully reimplement serde.

Is this actually a problem space we can shrink at all? 
Is there a way to have a smaller implementation that still provides enough benefit?

If the answer to the above is loosely "no", then are there other options you can look into before implementing a comprehensive deserialization library here?

I presume we can't really contribute upstream to serde as these would be pretty substantial breaking changes (and I'm unsure how easily we could implement fail-slow deserialization). Is this something we could maintain in a fork of `serde` instead?

Reply:
This is definetly a bit bigger than initially scoped, especially the `Visitor` trait and lifetime annotations.

Initially we actually had a simpler implementations, but after some considerations it turned out that those came with too many drawbacks regarding flexibility/allocations.

I think that this is still very much in scope, but it is something we should actively monitor. Our problem space is still (and hopefully will stay) small. We primarily rely on serde (or other backends like `rmp` or `bincode`) to do most of the heavy lifting.

I also think we mustn't succumb to the sunk-cost fallacy and actively monitor if a fork of serde, or another approach, might be more feasible. Right now, I'd say it is not. I think the development cost and inherited complexity of a fork would outweigh the benefits now right now.

Reply:
A contribution upstream is unlikely, as serde is very stable, as the fail-slow behavior does not align with its goals of serde. There have been some issues surrounding it, but due to its very stable nature and unlikely 2.0 release anytime soon/ever. I think any effort contributing is for naught and would also require quite a lot of restructuring (if one wanted to make use of the full `error-stack` capabilities)